File: Jeonghak/9251_LCS.cpp
#include<iostream>
#define MAX 1001

using namespace std;

string str1, str2;
int DP[MAX][MAX];

void find_LCS()
{
    for(int i=1; i<=str1.length(); i++){
        for(int j=1; j<=str2.length(); j++)
        {
            if(str1[i-1] == str2[j-1]) 
                DP[i][j] = DP[i-1][j-1] + 1;
            else
                DP[i][j] = max(DP[i-1][j], DP[i][j-1]);
        }
    }
}

int main()
{
    cin >> str1 >> str2;
    find_LCS();
    cout << DP[str1.length()][str2.length()];
}

File: hall_of_fame/ga-ng/9251_LCS.py
import sys

input = sys.stdin.readline

str1 = input().rstrip()
str2 = input().rstrip()

dp = [0] * max(len(str1), len(str2))

for i in range(len(str1)):
    cnt = 0
    for j in range(len(str2)):
        
        if cnt < dp[j]:
            cnt = dp[j]
        elif str1[i] == str2[j]:
            dp[j] = cnt + 1
            
print(max(dp))